multi page to single page conversion for own ca and ssr setup guides (#2643)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/sign-your-own-saml-csr/main/index.md

---
title: Sign the Okta certificate with your own CA
excerpt: How to use a custom SAML certificate for apps
layout: Guides
---

This guide teaches you how to upload your own SAML certificates to sign the assertion for Outbound SAML apps and to sign the AuthN request and decrypt the assertion for Inbound SAML.

---

**Learning outcomes**

* List your apps using Okta APIs.
* Generate, sign, and publish a certificate signing request (CSR).

> * Update your app, and ISV, to use the new certificate.

Reviewer: ```suggestion
* Update your app and ISV to use the new certificate.
```

* Clone the certificate for use across multiple apps.

**What you need**

> * To use your own SAML certificate, update the key credential for the affected apps or IdPs.

Reviewer: ```suggestion
* To use your SAML certificate, an updated key credential for the affected apps or IdPs
```

**Sample code**

n/a

---

## Outbound and Inbound SAML Applications

> Okta as a SAML Service Provider is referred to as Inbound SAML. Okta as a SAML Identity Provider (IdP) is referred to as Outbound SAML. The general procedure is the same for both, however, some of the API calls are different as described in the sections below. 

Reviewer: ```suggestion
Okta as a SAML Service Provider is referred to as the Inbound SAML. Okta as a SAML Identity Provider (IdP) is referred to as the Outbound SAML. The general procedure is the same for both, however, some of the API calls are different as described in the following sections. 
```

> > **Note:** After you update the key credential, your users can't access the SAML app until you upload the new certificate to the ISV.

Reviewer: ```suggestion
> **Note:** After you update the key credential, users can't access the SAML app until you upload the new certificate to the ISV.
```

> For information on using the Postman REST API test client for these steps, see [Get Started with the Okta REST APIs](/code/rest/).

Reviewer: ```suggestion
See [Get Started with the Okta REST APIs](/code/rest/) for information on how to use the Postman REST API test client for these steps.
```

## List your apps

> * For Outbound SAML, use the [Apps API](/docs/reference/api/apps/#list-applications) to return a list of all apps and collect the app `id` for each app that you want to update.

Reviewer: ```suggestion
* For Outbound SAML, use the [Apps API](/docs/reference/api/apps/#list-applications) to return a list of all the apps and to collect the app `id` for each app that you want to update.
```

    "status": "ACTIVE",
    "lastUpdated": "2015-01-24T00:09:01.000Z",
    "created": "2014-01-06T23:42:40.000Z",
    "accessibility": {
      "selfService": false,
      "errorRedirectUrl": null,
      "loginRedirectUrl": null
    },
    "visibility": {
      "autoSubmitToolbar": true,
      "hide": {
        "iOS": false,
        "web": false
      },
      "appLinks": {
        "login": true
      }
    },
    "features": [],
    "signOnMode": "SAML_2_0",
    "credentials": {
      "scheme": "EDIT_USERNAME_AND_PASSWORD",
      "userNameTemplate": {
        "template": "${source.login}",
        "type": "BUILT_IN"
      },
      "revealPassword": true,
      "signing": {
        "kid": "ZcLGUsl4Xn3996YYel6KPvOxZOhNWfly5-q36CByH4o"
      }
    },
    "settings": {
      "app": {
        "instanceType": null
      },
      "notifications": {
        "vpn": {
          "network": {
            "connection": "DISABLED"
          },
          "message": null,
          "helpUrl": null
        }
      }
    }
  }
```

## Generate a CSR

* Use the [Apps API](/docs/reference/api/apps/#list-csrs-for-application) to return a list of all apps to use with Outbound SAML apps.

> * Use the [IdPs API](/docs/reference/api/idps/#list-signing-csrs-for-idp) to return a list of all IdPs to use with Inbound SAML apps.

Reviewer: Suggest that you don't put these two items in a bullet list. Just a paragraph will do. Reason is that you don't have an intro sentence, so to start this section off with two loose bullet items is a bit awkward and doesn't read well.

      },
      "response": {
        "signature": {
          "algorithm": "SHA-256",
          "scope": "ANY"
        }
      }
    },
    "settings": {
      "nameFormat": "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"
    },
    "credentials": {
      "trust": {
        "issuer": "https://idp.example.com",
        "audience": "https://www.okta.com/saml2/service-provider/spCQJRNaaxs7ANqKBO7M",
        "kid": "ZcLGUsl4Xn3996YYel6KPvOxZOhNWfly5-q36CByH4o"
      },
      "signing": {
        "kid": "your-key-id"
      }
    }
  },
  "policy": {
    "provisioning": {
      "action": "AUTO",
      "profileMaster": true,
      "groups": {
        "action": "NONE"
      }
    },
    "accountLink": {
      "filter": null,
      "action": "AUTO"
    },
    "subject": {
      "userNameTemplate": {
        "template": "idpuser.subjectNameId"
      },
      "filter": null,
      "matchType": "USERNAME"
    },
    "maxClockSkew": 120000
  }
}
```

## Clone the certificate (optional)

* To share the certificate that you created across multiple apps, clone it with the
[Apps API](/docs/reference/api/apps/#clone-application-key-credential) using the key `id` that you generated.

> * To share the certificate that you created across multiple IdPs, clone it with the [IdPs API](#top) using the key `id` that you generated.

Reviewer: ```suggestion
To share the certificate that you created across multiple IdPs, clone it with the [IdPs API](#top) using the key `id` that you generated.
```

Be sure to clone the certificate to every app or IdP with which you want to share it. Once cloned, you will need to [update the key credential](/docs/guides/sign-your-own-saml-csr/update-the-key-credential/) for the target app.

If the certificate that you cloned from changes, you must repeat the cloning operation for each app or IdP.

> **Important:** Sharing certificates isn't a recommended security practice. This API is provided to support legacy use cases.

## Upload the new certificate to the ISV

> > **Caution:** After completing the [Publish a CSR with a certificate](#publish-a-CSR-with-a-certificate) step, your users can't access the SAML app or the Identity Provider until you complete this step.

Reviewer: ```suggestion
> **Caution:** After you complete the [Publish a CSR with a certificate](#publish-a-CSR-with-a-certificate) step, users can't access the SAML app or the Identity Provider until you complete this step.
```

For Outbound SAML, complete the following steps (note that these steps can't be automated):

1. In the Admin Console, go to **Applications** > **Applications**.
1. Select your app integration.
1. Select **Sign-On Options**.
1. Click **View Setup Instructions**, as shown below.<br/>![Accessing SAML Setup Instructions](/img/saml_setup_link.png)
1. Perform the setup for your app integration again by using the instructions provided. During this setup, you can upload the certificate in a specified format, the metadata, or the certificate fingerprint.

For Inbound SAML, follow the existing procedures for your setup.

> ## See also

Reviewer: ```suggestion
See [Get Started with the Okta REST APIs](/code/rest/) for more information.
```

> * [Get Started with the Okta REST APIs](/code/rest/)

Reviewer: Remove this line.
